File: brain/cli/registry.py
"""
Registry central para gestión de comandos.
"""
from typing import Dict, List
from brain.cli.base import BaseCommand
from brain.cli.categories import CommandCategory


class CommandRegistry:
    """Registry central con capacidades de introspección"""
    
    def __init__(self):
        self._commands: Dict[str, BaseCommand] = {}
        self._by_category: Dict[CommandCategory, List[BaseCommand]] = {}
    
    def register(self, command: BaseCommand) -> None:
        """Registra un comando y lo indexa por categoría"""
        meta = command.metadata()
        self._commands[meta.name] = command
        
        if meta.category not in self._by_category:
            self._by_category[meta.category] = []
        self._by_category[meta.category].append(command)
    
    def get_by_category(self, category: CommandCategory) -> List[BaseCommand]:
        """Obtiene todos los comandos de una categoría"""
        return self._by_category.get(category, [])
    
    def list_all(self) -> Dict[CommandCategory, List[str]]:
        """Lista todos los comandos agrupados por categoría"""
        return {
            cat: [cmd.metadata().name for cmd in cmds]
            for cat, cmds in self._by_category.items()
        }
    
    def get_all_commands(self) -> List[BaseCommand]:
        """Retorna todos los comandos registrados"""
        return list(self._commands.values())


# Singleton global
_registry = CommandRegistry()


def get_registry() -> CommandRegistry:
    """Retorna la instancia del registry"""
    return _registry

File: installer/metamorph/internal/core/registry.go
package core

import (
	"github.com/spf13/cobra"
)

// CommandFactory crea un comando usando Core
type CommandFactory func(*Core) *cobra.Command

// registry global de comandos por categoría
var commandRegistry = make(map[string][]CommandFactory)

// RegisterCommand registra un comando en una categoría
func RegisterCommand(category string, factory CommandFactory) {
	commandRegistry[category] = append(commandRegistry[category], factory)
}

// BuildCommands construye todos los comandos registrados
func BuildCommands(c *Core, root *cobra.Command) {
	for _, factories := range commandRegistry {
		for _, factory := range factories {
			cmd := factory(c)
			if cmd != nil {
				root.AddCommand(cmd)
			}
		}
	}
}
